In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import sklearn

In [2]:
xls=pd.ExcelFile('Customer Raw data.xlsx')

In [3]:
xls.sheet_names

['Product Description', 'Product Performance', 'Unbxd Query performance']

# importing sheets from excel file


In [4]:
PD=pd.read_excel(xls,sheet_name='Product Description')
PP=pd.read_excel(xls,sheet_name='Product Performance')
QP=pd.read_excel(xls,sheet_name='Unbxd Query performance')

C:\Users\rupes\anaconda3\lib\site-packages\openpyxl\worksheet\_reader.py:312: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


In [6]:
PD.head()

PID                                   Product Name Category  \
0  CL0036386            Ikat stripe print cotton knit dress    Dress   
1  CL0030567                    Feminine pleated knit dress    Dress   
2  CL0032320   Floral embellished chambray denim maxi dress    Dress   
3  CL0036700                     Dot print cap sleeve shirt      Top   
4  CL0036365          Layered asymmetric hem hacci knit top      Top   

   Final Price  Initial Price             Color Primary Color           Fabric  
0        84.95          33.95   Blue/navy/white          Blue   Cotton Spandex  
1        59.95          41.95         Navy blue     Navy blue   Cotton Spandex  
2        96.95          67.95       Deep indigo   Deep indigo           Cotton  
3        49.95          19.95        White/navy         White           Cotton  
4        69.95          27.95              Blue          Blue        Polyester

In [9]:
PD.shape

(716, 8)

In [7]:
PP.head()

pId                                  pName  clicks  carts  orders
0  CL0034516   Chevron stripe colorblock maxi dress   29144   2181     204
1  CL0037085   Chevron stripe colorblock maxi dress   25892    237      70
2  CL0036758   Bird print pleated cotton knit dress   16343    340      97
3  CL0032310        Cotton knit fit and flare dress   16234    350      99
4  CL0037110  Pleat front chambray denim maxi dress   13317    229      56

In [10]:
PP.shape

(2501, 5)

In [8]:
QP.head()

query  hits  clicks  clickThroughs    ctr  carts  \
0                       pink   3971     330             93   2.34     18   
1                    Chambray   612    1360            447  73.04    145   
2  gingham in green and white   601       0              0   0.00      0   
3               white dresses   486     800            310  63.79     73   
4                        grey   480      89             28   5.83     10   

   orders  
0       0  
1       8  
2       0  
3       1  
4       0

In [11]:
QP.shape

(500, 7)

# Checking number of unique product ids in production description and product performance

In [16]:
PD['PID'].nunique()

716

In [14]:
PP['pId'].nunique()

2501

In [90]:
PP['pName'].nunique()

1081

In [91]:
PD['Product Name'].nunique()

646

# Result below shows that all product IDs that are in product description also are in product performance. Also Product performance have more product IDs than mentioned in product description

In [19]:
len(list(set(PD['PID'].unique()).intersection(set(PP['pId'].unique()))))

716

# Taking info from Product performance and updating it into Product description according to the respective product ID.

In [22]:
for i in PP.columns[1:5]:
    PD[i]=None

In [34]:
for i in PP.columns[1:5]:
    for j in PD['PID']:
        PD.loc[PD['PID']==j,i]=PP.loc[PP['pId']==j,i].values[0]

In [37]:
PD.drop(columns=['pName'],inplace=True)

In [38]:
PD

PID                                   Product Name Category  \
0    CL0036386            Ikat stripe print cotton knit dress    Dress   
1    CL0030567                    Feminine pleated knit dress    Dress   
2    CL0032320   Floral embellished chambray denim maxi dress    Dress   
3    CL0036700                     Dot print cap sleeve shirt      Top   
4    CL0036365          Layered asymmetric hem hacci knit top      Top   
..         ...                                            ...      ...   
711  CL0031288                      Ruffle front poplin dress    Dress   
712  CL0032199              Beaded cotton poplin sheath dress    Dress   
713  CL0031267          Embellished stripe waist poplin dress    Dress   
714  CL0031287                        Sash waist poplin skirt    Skirt   
715  CL0031947                  Colorblock retro poplin frock    Dress   

     Final Price  Initial Price                Color        Primary Color  \
0          84.95          33.95      Blue/navy/white                 Blue   
1          59.95          41.95            Navy blue            Navy blue   
2          96.95          67.95          Deep indigo          Deep indigo   
3          49.95          19.95           White/navy                White   
4          69.95          27.95                 Blue                 Blue   
..           ...            ...                  ...                  ...   
711        69.95          52.95         Spring green         Spring green   
712        86.95          65.95         Spring green         Spring green   
713        76.95          57.95   Spring green multi   Spring green multi   
714        59.95          44.95         Spring green         Spring green   
715        59.95          44.95   Black/spring green                Black   

              Fabric clicks carts orders  
0     Cotton Spandex   1846    43     15  
1     Cotton Spandex   6882   242     56  
2             Cotton  10680    72     27  
3             Cotton   2051    59     24  
4          Polyester   4630   123     50  
..               ...    ...   ...    ...  
711           Cotton     62     8      0  
712           Cotton    380     4      0  
713           Cotton    175    16      2  
714           Cotton    108    15      0  
715           Cotton   2774   102     24  

[716 rows x 11 columns]

# changing type of clicks,carts and orders columns to integer type.

In [56]:
for i in ['clicks','carts','orders']:
    PD[i]=PD[i].astype('int32')

In [57]:
PD.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 716 entries, 0 to 715
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PID            716 non-null    object 
 1   Product Name   716 non-null    object 
 2   Category       716 non-null    object 
 3   Final Price    716 non-null    float64
 4   Initial Price  716 non-null    float64
 5   Color          716 non-null    object 
 6   Primary Color  716 non-null    object 
 7   Fabric         716 non-null    object 
 8   clicks         716 non-null    int32  
 9   carts          716 non-null    int32  
 10  orders         716 non-null    int32  
dtypes: float64(2), int32(3), object(6)
memory usage: 53.3+ KB


# Grouping product data on the basis of colours, Fabric and Category for exploration only

In [164]:
color=PD.groupby(['Primary Color'])['clicks','carts','orders'].agg(['mean','count'])

<ipython-input-164-48e5e0b1c0b1>:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  color=PD.groupby(['Primary Color'])['clicks','carts','orders'].agg(['mean','count'])


# PRIMARY COLOR

In [165]:
color

clicks            carts           orders      
                          mean count       mean count       mean count
Primary Color                                                         
 Aqua               488.714286     7   9.571429     7   2.142857     7
 Aqua green multi  1878.000000     1  22.000000     1  13.000000     1
 Aqua multi        2309.000000     2  25.000000     2  10.500000     2
 Banana yellow      777.333333     3  42.000000     3   3.666667     3
 Beige              734.909091    11  11.272727    11   4.090909    11
...                        ...   ...        ...   ...        ...   ...
 Tomato red         755.500000     2  36.000000     2  11.500000     2
 Turquoise          949.333333     3  16.333333     3   3.666667     3
 White             1307.355932    59  26.322034    59   7.254237    59
 White multi       1452.000000     3  35.000000     3  10.000000     3
 Yellow             680.181818    11  15.545455    11   3.818182    11

[122 rows x 6 columns]

# sorting on average no. of clicks

In [166]:
color.sort_values(by=('clicks','mean'),ascending=False)

clicks            carts           orders      
                            mean count       mean count       mean count
Primary Color                                                           
 Light blue          4981.250000     4  98.250000     4  28.750000     4
 Purple              4279.000000     3  52.333333     3  18.333333     3
 Indigo              3901.916667    12  92.916667    12  25.666667    12
 Indigo blue         3666.000000     1  99.000000     1  20.000000     1
 Peach               2917.000000     1  65.000000     1  14.000000     1
...                          ...   ...        ...   ...        ...   ...
 Lavender              70.000000     2   3.000000     2   2.500000     2
 Stellar blue          55.000000     1   2.000000     1   2.000000     1
 Crimson red           52.000000     1   8.000000     1   2.000000     1
 Light melange gray    40.000000     1   6.000000     1   0.000000     1
 Leaf green            39.000000     1   2.000000     1   0.000000     1

[122 rows x 6 columns]

# sorting on count of particular category in data

In [167]:
color.sort_values(by=('clicks','count'),ascending=False)

clicks            carts           orders      
                      mean count       mean count       mean count
Primary Color                                                     
 Black         1513.320000   100  32.860000   100   9.200000   100
 White         1307.355932    59  26.322034    59   7.254237    59
 Off-white     1174.321429    56  18.196429    56   5.035714    56
 Deep navy     1566.204545    44  42.977273    44  11.409091    44
 Deep indigo   1644.896552    29  40.551724    29   9.275862    29
...                    ...   ...        ...   ...        ...   ...
 Peach bloom   1902.000000     1  30.000000     1  15.000000     1
 Chocolate      203.000000     1   4.000000     1   2.000000     1
 Pink multi    1756.000000     1  15.000000     1   7.000000     1
 Indigo blue   3666.000000     1  99.000000     1  20.000000     1
 Powder blue    575.000000     1   9.000000     1   3.000000     1

[122 rows x 6 columns]

# FABRIC

# sorting on average no. of clicks


In [168]:
fabric=PD.groupby(['Fabric'])['clicks','carts','orders'].agg(['mean','count'])

<ipython-input-168-fb7b4bc57806>:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  fabric=PD.groupby(['Fabric'])['clicks','carts','orders'].agg(['mean','count'])


In [169]:
fabric.sort_values(by=('clicks','mean'),ascending=False)

clicks            carts           orders      
                                  mean count       mean count       mean count
Fabric                                                                        
 Nylon Polyester           2041.500000     2  27.000000     2  12.000000     2
 Cotton Nylon              1859.833333     6  29.333333     6   8.333333     6
 Cotton                    1695.372624   263  39.235741   263  10.365019   263
 Nylon Cotton              1395.000000     1   6.000000     1   0.000000     1
 Cotton Spandex            1291.978417   139  33.474820   139   9.877698   139
 Cotton Polyester          1260.941176    17  17.294118    17   4.411765    17
 Silk                      1215.142857     7   2.571429     7   0.142857     7
 Polyester Nylon            983.250000     4  10.000000     4   4.750000     4
 Cotton Linen               979.000000     1  20.000000     1   6.000000     1
 Nylon                      976.666667     3   5.666667     3   1.000000     3
 Polyester Metallic         939.000000     2  19.000000     2   0.500000     2
 Polyester                  938.846154   104  15.009615   104   4.163462   104
 Polyester Rayon Spandex    897.000000     4  23.500000     4   5.500000     4
 Polyester Crepe de Chine   894.046512    43  15.325581    43   3.744186    43
 Tencel                     816.333333     3  10.666667     3   2.333333     3
 Cotton Rayon Spandex       771.000000     4   7.000000     4   3.000000     4
 Polyamide Spandex          745.333333    12  16.166667    12   4.833333    12
 Cotton Polyester Spandex   724.941176    17  13.588235    17   4.117647    17
 Rayon Polyester Spandex    676.285714     7  16.571429     7   4.428571     7
 Viscose                    638.250000     4  14.500000     4   5.250000     4
 Viscose Spandex            600.550000    20  12.250000    20   3.200000    20
 Rayon Nylon Spandex        564.083333    12  14.833333    12   3.750000    12
 Rayon                      558.166667     6   8.666667     6   2.666667     6
 Polyester Spandex          543.468750    32   9.843750    32   2.718750    32
 Polyester Rayon            331.000000     2   9.500000     2   2.000000     2
 Wool Nylon                 227.000000     1   4.000000     1   0.000000     1

# sorting on count of particular category in data

In [170]:
fabric.sort_values(by=('clicks','count'),ascending=False)

clicks            carts           orders      
                                  mean count       mean count       mean count
Fabric                                                                        
 Cotton                    1695.372624   263  39.235741   263  10.365019   263
 Cotton Spandex            1291.978417   139  33.474820   139   9.877698   139
 Polyester                  938.846154   104  15.009615   104   4.163462   104
 Polyester Crepe de Chine   894.046512    43  15.325581    43   3.744186    43
 Polyester Spandex          543.468750    32   9.843750    32   2.718750    32
 Viscose Spandex            600.550000    20  12.250000    20   3.200000    20
 Cotton Polyester          1260.941176    17  17.294118    17   4.411765    17
 Cotton Polyester Spandex   724.941176    17  13.588235    17   4.117647    17
 Polyamide Spandex          745.333333    12  16.166667    12   4.833333    12
 Rayon Nylon Spandex        564.083333    12  14.833333    12   3.750000    12
 Rayon Polyester Spandex    676.285714     7  16.571429     7   4.428571     7
 Silk                      1215.142857     7   2.571429     7   0.142857     7
 Cotton Nylon              1859.833333     6  29.333333     6   8.333333     6
 Rayon                      558.166667     6   8.666667     6   2.666667     6
 Viscose                    638.250000     4  14.500000     4   5.250000     4
 Cotton Rayon Spandex       771.000000     4   7.000000     4   3.000000     4
 Polyester Nylon            983.250000     4  10.000000     4   4.750000     4
 Polyester Rayon Spandex    897.000000     4  23.500000     4   5.500000     4
 Nylon                      976.666667     3   5.666667     3   1.000000     3
 Tencel                     816.333333     3  10.666667     3   2.333333     3
 Polyester Metallic         939.000000     2  19.000000     2   0.500000     2
 Polyester Rayon            331.000000     2   9.500000     2   2.000000     2
 Nylon Polyester           2041.500000     2  27.000000     2  12.000000     2
 Cotton Linen               979.000000     1  20.000000     1   6.000000     1
 Nylon Cotton              1395.000000     1   6.000000     1   0.000000     1
 Wool Nylon                 227.000000     1   4.000000     1   0.000000     1

# Category

# sorting on average no. of clicks


In [171]:
categ=PD.groupby(['Category'])['clicks','carts','orders'].agg(['mean','count'])

<ipython-input-171-145ea1e54a84>:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  categ=PD.groupby(['Category'])['clicks','carts','orders'].agg(['mean','count'])


In [172]:
categ.sort_values(by=('clicks','mean'),ascending=False)

clicks            carts          orders      
                  mean count       mean count      mean count
Category                                                     
 Dress     1678.658879   428  34.864486   428  9.324766   428
 Jumpsuit  1000.545455    11  13.454545    11  2.818182    11
 Skirt      771.083333    72  22.083333    72  5.500000    72
 Jacket     640.888889     9  17.555556     9  4.666667     9
 Top        557.877551   147  14.653061   147  4.768707   147
 Pant       472.857143    49   9.163265    49  3.020408    49

# sorting on count of particular category in data

In [173]:
categ.sort_values(by=('clicks','count'),ascending=False)

clicks            carts          orders      
                  mean count       mean count      mean count
Category                                                     
 Dress     1678.658879   428  34.864486   428  9.324766   428
 Top        557.877551   147  14.653061   147  4.768707   147
 Skirt      771.083333    72  22.083333    72  5.500000    72
 Pant       472.857143    49   9.163265    49  3.020408    49
 Jumpsuit  1000.545455    11  13.454545    11  2.818182    11
 Jacket     640.888889     9  17.555556     9  4.666667     9

# ------

# Lets look at the query performance

In [132]:
 QP.head(n=15)

query  hits  clicks  clickThroughs    ctr  carts  \
0                        pink   3971     330             93   2.34     18   
1                     Chambray   612    1360            447  73.04    145   
2   gingham in green and white   601       0              0   0.00      0   
3                white dresses   486     800            310  63.79     73   
4                         grey   480      89             28   5.83     10   
5                   grey dress   435      65             29   6.67      7   
6                        ,maxi   378     888            286  75.66    117   
7                       green    376     723            237  63.03     64   
8                        birds   350     626            235  67.14     78   
9                       Purple   340     416            151  44.41     29   
10                        Reds   332     621            151  45.48     68   
11                Maxi dresses   327     633            237  72.48     30   
12               black dresses   303     655            233  76.90     42   
13                    Chevron    284     449            223  78.52     72   
14                      white    282     610            193  68.44     67   

    orders  
0        0  
1        8  
2        0  
3        1  
4        0  
5        0  
6       13  
7        9  
8        4  
9        0  
10       1  
11       1  
12       2  
13       7  
14       1

# we can see that for some queries , the ctr is very less, lets take some cases

In [149]:
low_ctr=QP[QP['ctr']<20]

In [150]:
low_ctr

query  hits  clicks  \
0                                                pink   3971     330   
2                           gingham in green and white   601       0   
4                                                 grey   480      89   
5                                           grey dress   435      65   
50                                        Teal dresses   122      25   
55                                  black lace dresses   108       5   
83                                               coats    73      94   
139                                 Havana dress-Black    45      29   
174                                              swim     37      24   
238  Paisley embellished voile tunic dress-White multi    27       5   
253                                           referral    26       7   
288                             Ellen dress-Aqua multi    23      31   
290            Floral gold lace A-line dress-Ecru/gold    23      17   
302      Empire waist stretch ottoman dress-Teal green    22      15   
311                                      return policy    21      17   
398                                          umbrellas    17      34   
419       Elastic waist gingham maxi dress-Black/white    16       9   
432                                              Beige    16      10   
466  Polka dot chiffon surplice dress-Deep navy/orange    15       4   
468                                            Referal    14      28   
470                                             Kisses    14       5   
474  Belted stripe poplin dress-White/juniper green...    14       2   
475     Corset style cotton poplin dress-Prussian blue    14       3   

     clickThroughs    ctr  carts  orders  
0               93   2.34     18       0  
2                0   0.00      0       0  
4               28   5.83     10       0  
5               29   6.67      7       0  
50              13  10.66      2       0  
55               4   3.70      2       0  
83              12  16.44      2       0  
139              7  15.56      3       0  
174              7  18.92      2       1  
238              3  11.11      0       0  
253              3  11.54      0       0  
288              4  17.39     22       0  
290              4  17.39      4       0  
302              4  18.18      4       1  
311              3  14.29      2       0  
398              3  17.65      0       0  
419              1   6.25      0       0  
432              3  18.75      0       0  
466              2  13.33      0       0  
468              2  14.29      8       1  
470              2  14.29      8       0  
474              2  14.29      2       1  
475              2  14.29      0       0

# lets look at the first query:'pink '
1. we can grab product with pink colour from product description
2. Then we can compare the result.
3. Ater comparing we can try to figure out where our search product is lacking

In [136]:
df=PD.groupby(['Primary Color'])['clicks','carts','orders'].sum()

<ipython-input-136-b20846e314b4>:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df=PD.groupby(['Primary Color'])['clicks','carts','orders'].sum()


In [137]:
df

clicks  carts  orders
Primary Color                           
 Aqua                3421     67      15
 Aqua green multi    1878     22      13
 Aqua multi          4618     50      21
 Banana yellow       2332    126      11
 Beige               8084    124      45
...                   ...    ...     ...
 Tomato red          1511     72      23
 Turquoise           2848     49      11
 White              77134   1553     428
 White multi         4356    105      30
 Yellow              7482    171      42

[122 rows x 3 columns]

In [138]:
df.index

Index([' Aqua', ' Aqua green multi', ' Aqua multi', ' Banana yellow', ' Beige',
       ' Beige multi', ' Black', ' Black multi', ' Blue', ' Blue multi',
       ...
       ' Surf blue', ' Teal', ' Teal green', ' Teal multi', ' Tiger lily',
       ' Tomato red', ' Turquoise', ' White', ' White multi', ' Yellow'],
      dtype='object', name='Primary Color', length=122)

In [141]:
df.loc[' Pink',:]

clicks    1803
carts       46
orders       7
Name:  Pink, dtype: int32

In [148]:
QP.loc[QP['query']=='pink ',:]

query  hits  clicks  clickThroughs   ctr  carts  orders
0  pink   3971     330             93  2.34     18       0

# What I observed
1. we can see that no. of hits(3971) is quite good but that is not converting into clicks(330) then inturn into click throughs(93) thats producing a low ctr of 2.34%
2. On the other hand, pink products are gaining quite a number of clicks(1803), So people are reaching pink products not through the query 'pink ' but from somewhere else only and also buying them.
3. Above statement clearly shows that people want pink products but they are not getting the query results they need from search product.
-- Hence there may be some modifiction that are needed in the search product to produce better result for query 'pink '  

# Lets take another case of low ctr. 

In [151]:
low_ctr.head(3)

query  hits  clicks  clickThroughs   ctr  carts  \
0                       pink   3971     330             93  2.34     18   
2  gingham in green and white   601       0              0  0.00      0   
4                        grey   480      89             28  5.83     10   

   orders  
0       0  
2       0  
4       0

The second case of low ctr query 'gingham in green and white ' is getting reasonable amount of hits(601) but 0 clicks. That is quite shocking means it is only possible if the search product does not produce any result for this query
-- lets search for this type of product in product description

In [153]:
PD[PD['Product Name'].str.contains('gingham|Gingham')]

PID                               Product Name Category  \
88   CL0035968      Split neck gingham check sheath dress    Dress   
146  CL0036621               Bow tied gingham check dress    Dress   
187  CL0035880     Gingham check cotton belted shirtdress    Dress   
203  CL0036336      Embellished gingham check shift dress    Dress   
367  CL0035520                Cotton gingham check shorts     Pant   
589  CL0036846   Flutter sleeve gingham check shift dress    Dress   
686  CL0037096           Elastic waist gingham maxi dress    Dress   

     Final Price  Initial Price                  Color Primary Color  \
88         64.95          42.95   Mint green/off-white    Mint green   
146       101.95          50.95         Cream/red/teal         Cream   
187        69.95          45.95        Black/off-white         Black   
203        96.95          58.95           Yellow/white        Yellow   
367        79.95          47.95            Black/white         Black   
589        59.95          38.95            Black/white         Black   
686        69.95          52.95            Black/white         Black   

              Fabric  clicks  carts  orders  
88    Cotton Spandex     631     11       6  
146           Cotton     889     15       4  
187           Cotton    1601     40      11  
203           Cotton    1176     28       6  
367           Cotton     310      6       0  
589     Cotton Linen     979     20       6  
686           Cotton    6848    103      42

# We can see that the first result with index 88 can be useful to customers searching for 'gingham in green and white ' but the search product must not be showing this product thus resulting in 0 clicks.

Suggestion: search product could use a better algorithm to seperate out the words in search query and find it in categories of product description.

In [154]:
low_ctr

query  hits  clicks  \
0                                                pink   3971     330   
2                           gingham in green and white   601       0   
4                                                 grey   480      89   
5                                           grey dress   435      65   
50                                        Teal dresses   122      25   
55                                  black lace dresses   108       5   
83                                               coats    73      94   
139                                 Havana dress-Black    45      29   
174                                              swim     37      24   
238  Paisley embellished voile tunic dress-White multi    27       5   
253                                           referral    26       7   
288                             Ellen dress-Aqua multi    23      31   
290            Floral gold lace A-line dress-Ecru/gold    23      17   
302      Empire waist stretch ottoman dress-Teal green    22      15   
311                                      return policy    21      17   
398                                          umbrellas    17      34   
419       Elastic waist gingham maxi dress-Black/white    16       9   
432                                              Beige    16      10   
466  Polka dot chiffon surplice dress-Deep navy/orange    15       4   
468                                            Referal    14      28   
470                                             Kisses    14       5   
474  Belted stripe poplin dress-White/juniper green...    14       2   
475     Corset style cotton poplin dress-Prussian blue    14       3   

     clickThroughs    ctr  carts  orders  
0               93   2.34     18       0  
2                0   0.00      0       0  
4               28   5.83     10       0  
5               29   6.67      7       0  
50              13  10.66      2       0  
55               4   3.70      2       0  
83              12  16.44      2       0  
139              7  15.56      3       0  
174              7  18.92      2       1  
238              3  11.11      0       0  
253              3  11.54      0       0  
288              4  17.39     22       0  
290              4  17.39      4       0  
302              4  18.18      4       1  
311              3  14.29      2       0  
398              3  17.65      0       0  
419              1   6.25      0       0  
432              3  18.75      0       0  
466              2  13.33      0       0  
468              2  14.29      8       1  
470              2  14.29      8       0  
474              2  14.29      2       1  
475              2  14.29      0       0

# Lets take a look at query grey and grey dress producing a ctr of 5.8 and 6.3% respectively

No result for colr grey or primary color grey in product description

In [160]:
PD[PD['Color'].str.contains('Grey|grey')]

Empty DataFrame
Columns: [PID, Product Name, Category, Final Price, Initial Price, Color, Primary Color, Fabric, clicks, carts, orders]
Index: []

In [161]:
PD[PD['Primary Color'].str.contains('Grey|grey')]

Empty DataFrame
Columns: [PID, Product Name, Category, Final Price, Initial Price, Color, Primary Color, Fabric, clicks, carts, orders]
Index: []

But there are quiet a number of results for Black/white and White/black  which can be grey and can be useful to customers buying clothes. The Companies may not have grey color in Primary Color yet.  

In [162]:
PD[PD['Color'].str.contains('Black/white|White/black')]

PID                                   Product Name Category  \
9    CL0036375                Balloon print cotton knit dress    Dress   
33   CL0036708         Contrast trim balloon print maxi dress    Dress   
85   CL0035685            Piped trim floral print voile dress    Dress   
94   CL0035928            Jacquard check banded collar blazer   Jacket   
154  CL0036372   Contrast seersucker stripe poplin shirtdress    Dress   
170  CL0036064       Graphic stripe crepe elastic waist pants     Pant   
172  CL0036235      Floral embellished split neck voile dress    Dress   
190  CL0036796           Chevron stripe colorblock maxi dress    Dress   
200  CL0035630               Elastic belted stripe knit dress    Dress   
202  CL0034414                   Her fifties piped trim dress    Dress   
209  CL0035736                 Stripe jersey knit shift dress    Dress   
215  CL0035622                 Colorblock stripe poplin dress    Dress   
238  CL0035797                Trees digital print shift dress    Dress   
241  CL0035804              Banded stripe cotton poplin dress    Dress   
262  CL0035494              Stretch jacquard check crop pants     Pant   
272  CL0036408            Stripe layered asymmetric hem dress    Dress   
291  CL0035937         Floral neck jersey stripe sheath dress    Dress   
332  CL0034625                 Piped trim cotton poplin dress    Dress   
348  CL0035803             Chevron banded stripe poplin dress    Dress   
363  CL0036130                 Emoji embellished jogger pants     Pant   
367  CL0035520                    Cotton gingham check shorts     Pant   
376  CL0034592                  Stripe trim cotton knit tunic      Top   
383  CL0036427            Floral print cotton knit wrap dress    Dress   
429  CL0035214            Floral embellished voile caftan top      Top   
447  CL0036227                      Draped kimono knit jacket   Jacket   
497  CL0035845               Stripe voile dolman sleeve tunic      Top   
525  CL0036627            Scallop applique banded retro frock    Dress   
546  CL0035958                 Two tone check print crepe top      Top   
551  CL0036584           Chequered hat print split neck tunic      Top   
568  CL0036884          Stripe chiffon print angled hem dress    Dress   
589  CL0036846       Flutter sleeve gingham check shift dress    Dress   
615  CL0037019             Bold stripe cold shoulder knit top      Top   
653  CL0037085           Chevron stripe colorblock maxi dress    Dress   
686  CL0037096               Elastic waist gingham maxi dress    Dress   

     Final Price  Initial Price                  Color Primary Color  \
9          69.95          27.95            White/black         White   
33         79.95          31.95            White/black         White   
85         89.95          58.95     Black/white/yellow         Black   
94         99.95          59.95            Black/white         Black   
154        84.95          33.95            White/black         White   
170        79.95          51.95            Black/white         Black   
172        86.95          43.95            White/black         White   
190        99.95          79.95            White/black         White   
200       109.95          65.95            Black/white         Black   
202        79.95          51.95            White/black         White   
209        79.95          51.95            Black/white         Black   
215        79.95          55.95            Black/white         Black   
238        79.95          51.95       White/black/aqua         White   
241        79.95          39.95            White/black         White   
262        89.95          58.95            Black/white         Black   
272       119.95          47.95            Black/white         Black   
291        96.95          58.95            White/black         White   
332        89.95          53.95            White/black         White   
348        69.95          34.95            White/bla

The search product may be missing many of the above products due lack of color grey in product description.

# As we can see the low ctr for some queries bring out some reasonable problem in search product.

# ---------

# Further suggestion on the basis of Product Description and Product Performance

In [174]:
color

clicks            carts           orders      
                          mean count       mean count       mean count
Primary Color                                                         
 Aqua               488.714286     7   9.571429     7   2.142857     7
 Aqua green multi  1878.000000     1  22.000000     1  13.000000     1
 Aqua multi        2309.000000     2  25.000000     2  10.500000     2
 Banana yellow      777.333333     3  42.000000     3   3.666667     3
 Beige              734.909091    11  11.272727    11   4.090909    11
...                        ...   ...        ...   ...        ...   ...
 Tomato red         755.500000     2  36.000000     2  11.500000     2
 Turquoise          949.333333     3  16.333333     3   3.666667     3
 White             1307.355932    59  26.322034    59   7.254237    59
 White multi       1452.000000     3  35.000000     3  10.000000     3
 Yellow             680.181818    11  15.545455    11   3.818182    11

[122 rows x 6 columns]

In [176]:
color[('clicks','mean')].describe()

count     122.000000
mean     1064.505086
std       837.487023
min        39.000000
25%       545.000000
50%       874.541667
75%      1320.213983
max      4981.250000
Name: (clicks, mean), dtype: float64

In [177]:
color[('clicks','count')].describe()

count    122.000000
mean       5.868852
std       12.519704
min        1.000000
25%        1.000000
50%        2.000000
75%        5.000000
max      100.000000
Name: (clicks, count), dtype: float64

In [182]:
color[(color[('clicks','mean')]>1320) & (color[('clicks','count')]>5)]

clicks            carts           orders      
                         mean count       mean count       mean count
Primary Color                                                        
 Black            1513.320000   100  32.860000   100   9.200000   100
 Deep indigo      1644.896552    29  40.551724    29   9.275862    29
 Deep navy        1566.204545    44  42.977273    44  11.409091    44
 Indigo           3901.916667    12  92.916667    12  25.666667    12
 Navy blue        2362.666667     6  85.166667     6  22.666667     6
 Off-white multi  1324.500000     8  13.750000     8   5.750000     8
 Poppy red        1811.666667     9  45.666667     9  17.555556     9
 Red              1723.142857    14  26.285714    14   8.071429    14

# TABLE 1 
colors with higher no. of clicks(>75% is taken as boundary) and there are more no. of items available to show to the customer(>75% of the count) 

In [184]:
color[(color[('clicks','mean')]<545) & (color[('clicks','count')]<2)]

clicks       carts       orders      
                      mean count  mean count   mean count
Primary Color                                            
 Blush pink          540.0     1  17.0     1    5.0     1
 Chocolate           203.0     1   4.0     1    2.0     1
 Crimson red          52.0     1   8.0     1    2.0     1
 Curry yellow        271.0     1   5.0     1    1.0     1
 Deep coral          185.0     1  10.0     1    2.0     1
 Deep gray           274.0     1  14.0     1    2.0     1
 Heather gray         96.0     1   7.0     1    1.0     1
 Juniper             534.0     1  34.0     1    6.0     1
 Kentucky blue       219.0     1   2.0     1    0.0     1
 Khaki brown         352.0     1   5.0     1    2.0     1
 Leaf green           39.0     1   2.0     1    0.0     1
 Light gray          243.0     1   4.0     1    2.0     1
 Light melange gray   40.0     1   6.0     1    0.0     1
 Lime                360.0     1   8.0     1    4.0     1
 Mandarin orange     189.0     1   6.0     1    0.0     1
 Mustard             227.0     1   4.0     1    0.0     1
 Prussian blue       470.0     1   2.0     1    0.0     1
 Rose smoke          167.0     1   0.0     1    3.0     1
 Smoke blue          206.0     1   8.0     1    3.0     1
 Spring green multi  175.0     1  16.0     1    2.0     1
 Stellar blue         55.0     1   2.0     1    2.0     1
 Teal multi          476.0     1   9.0     1    5.0     1
 Tiger lily          519.0     1   9.0     1    2.0     1

# TABLE 2
colors with lower no. of clicks(<25% is taken as boundary) and there are less no. of items available to show to the customer(<25% of the count) 

So for furthere suggestion, In a query like a sentence where 2 or more keywords are present where customer provide a choice, 
# Search Product should push contents of TABLE 1 first and try avoid pushing content of TABLE 2

# --------

Same can be done with fabric and category

# FABRIC

In [186]:
fabric[('clicks','mean')].describe()

count      26.000000
mean      944.768848
std       441.594082
min       227.000000
25%       647.758929
50%       895.523256
75%      1157.169643
max      2041.500000
Name: (clicks, mean), dtype: float64

In [187]:
fabric[('clicks','count')].describe()

count     26.000000
mean      27.538462
std       57.963596
min        1.000000
25%        3.000000
50%        6.000000
75%       17.000000
max      263.000000
Name: (clicks, count), dtype: float64

In [191]:
fabric[(fabric[('clicks','mean')]>1157) & (fabric[('clicks','count')]>16)]

clicks            carts           orders      
                          mean count       mean count       mean count
Fabric                                                                
 Cotton            1695.372624   263  39.235741   263  10.365019   263
 Cotton Polyester  1260.941176    17  17.294118    17   4.411765    17
 Cotton Spandex    1291.978417   139  33.474820   139   9.877698   139

# TABLE 1 
fabrics with higher no. of clicks(>75% is taken as boundary) and there are more no. of items available to show to the customer(>75% of the count) 

In [193]:
fabric[(fabric[('clicks','mean')]<647) & (fabric[('clicks','count')]<4)]

clicks       carts       orders      
                   mean count  mean count   mean count
Fabric                                                
 Polyester Rayon  331.0     2   9.5     2    2.0     2
 Wool Nylon       227.0     1   4.0     1    0.0     1

# TABLE 2
fabrics with lower no. of clicks(<25% is taken as boundary) and there are less no. of items available to show to the customer(<25% of the count) 

# CATEGORY

In [194]:
categ[('clicks','mean')].describe()

count       6.000000
mean      853.651875
std       444.249257
min       472.857143
25%       578.630385
50%       705.986111
75%       943.179924
max      1678.658879
Name: (clicks, mean), dtype: float64

In [195]:
categ[('clicks','count')].describe()

count      6.000000
mean     119.333333
std      159.459922
min        9.000000
25%       20.500000
50%       60.500000
75%      128.250000
max      428.000000
Name: (clicks, count), dtype: float64

In [200]:
categ[(categ[('clicks','mean')]>700) & (categ[('clicks','count')]>25)]

clicks            carts          orders      
                 mean count       mean count      mean count
Category                                                    
 Dress    1678.658879   428  34.864486   428  9.324766   428
 Skirt     771.083333    72  22.083333    72  5.500000    72

# TABLE 1 
clothes category with higher no. of clicks and there are more no. of items available to show to the customer

In [203]:
categ[(categ[('clicks','mean')]<578) & (categ[('clicks','count')]<50)]

clicks           carts          orders      
                mean count      mean count      mean count
Category                                                  
 Pant     472.857143    49  9.163265    49  3.020408    49

# TABLE 2
clothes category with lower no. of clicks and there are less no. of items available to show to the customer

# --------

In [205]:
PP.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2501 entries, 0 to 2500
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   pId     2501 non-null   object
 1   pName   2501 non-null   object
 2   clicks  2501 non-null   int64 
 3   carts   2501 non-null   int64 
 4   orders  2501 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 97.8+ KB


In [206]:
PP.corr()

clicks     carts    orders
clicks  1.000000  0.730533  0.877850
carts   0.730533  1.000000  0.801904
orders  0.877850  0.801904  1.000000

As we can see carts and orders have very high correlation with no. of clicks, we can safely focus on improving no. of clicks for search product analysis.

In [207]:
PP['orders/clicks']=PP['orders']/PP['clicks']

In [208]:
PP.head()

pId                                  pName  clicks  carts  orders  \
0  CL0034516   Chevron stripe colorblock maxi dress   29144   2181     204   
1  CL0037085   Chevron stripe colorblock maxi dress   25892    237      70   
2  CL0036758   Bird print pleated cotton knit dress   16343    340      97   
3  CL0032310        Cotton knit fit and flare dress   16234    350      99   
4  CL0037110  Pleat front chambray denim maxi dress   13317    229      56   

   orders/clicks  
0       0.007000  
1       0.002704  
2       0.005935  
3       0.006098  
4       0.004205

In [220]:
PP[PP['clicks']>100].sort_values(by=['orders/clicks'],ascending=False).head(8)

pId                                       pName  clicks  carts  \
1154  CL0036866  Pintuck pleat crinkle cotton trapeze dress     123      9   
807   CL0036433        Floral embellished poplin maxi dress     396     44   
1040  CL0032657             Cotton poplin A-line shirtdress     199     23   
926   CL0032311             Cotton knit fit and flare dress     279     30   
1133  CL0032925                   Tiered cotton voile tunic     133      4   
984   CL0037086        Chevron stripe colorblock maxi dress     236     24   
495   CL0037102        Chevron stripe colorblock maxi dress     857     46   
1010  CL0036660               Jersey knit stripe long shirt     217      8   

      orders  orders/clicks  
1154       7       0.056911  
807       21       0.053030  
1040      10       0.050251  
926       13       0.046595  
1133       6       0.045113  
984        9       0.038136  
495       29       0.033839  
1010       7       0.032258

# These could become highlights or default product listing as the orders/clicks is high as well as no. of clicks is high(meaning it is searched and clicked frequently)

# --------

Product performance contains many rows where name of the product is not available and i am guessing that's why their information is not present in product descriprtion

In [221]:
noname=PP[PP['pName']=='Name N/A']

In [222]:
noname

pId     pName  clicks  carts  orders  orders/clicks
790   CL0034516_OS  Name N/A     420      0      12       0.028571
799   CL0035152_OS  Name N/A     405      0       4       0.009877
886   CL0034663_OS  Name N/A     311      0       0       0.000000
904      CL0037242  Name N/A     300      4       1       0.003333
907   CL0032845_OS  Name N/A     297      0       5       0.016835
...            ...       ...     ...    ...     ...            ...
2495     CL0036125  Name N/A       0      0       3            inf
2496     CL0036376  Name N/A       0      0       1            inf
2498     GC0000202  Name N/A       0      0       5            inf
2499     GC0000204  Name N/A       0      0       7            inf
2500     GC0000205  Name N/A       0      0      11            inf

[1230 rows x 6 columns]

 -I left out these rows in the analysis where keywords were required for most of the analysis 
 -These could not be updated in product description as their PID is not present in product description.
 -Above data was included in calculation of correlation b/w carts,orders and clicks 

# ____ END ____